File: src/main/java/me/deecaad/weaponmechanicsplus/weapon/modifiers/ReloadModifier.java
package me.deecaad.weaponmechanicsplus.weapon.modifiers;

import me.deecaad.core.file.SerializeData;
import me.deecaad.core.file.Serializer;
import me.deecaad.core.file.SerializerException;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.IntegerModifier;
import org.jetbrains.annotations.NotNull;

public class ReloadModifier implements Serializer<ReloadModifier> {

    private IntegerModifier magazineSize;
    private IntegerModifier ammoPerReload;
    private IntegerModifier reloadDuration;
    private IntegerModifier shootDelayAfterReload;

    /**
     * Default constructor for serializer
     */
    public ReloadModifier() {
    }

    public ReloadModifier(IntegerModifier magazineSize, IntegerModifier ammoPerReload, IntegerModifier reloadDuration, IntegerModifier shootDelayAfterReload) {
        this.magazineSize = magazineSize;
        this.ammoPerReload = ammoPerReload;
        this.reloadDuration = reloadDuration;
        this.shootDelayAfterReload = shootDelayAfterReload;
    }

    @NotNull
    @Override
    public ReloadModifier serialize(SerializeData data) throws SerializerException {
        IntegerModifier magazineSize = data.of("Magazine_Size").serialize(IntegerModifier.class);
        IntegerModifier ammoPerReload = data.of("Ammo_Per_Reload").serialize(IntegerModifier.class);
        IntegerModifier reloadDuration = data.of("Reload_Duration").serialize(IntegerModifier.class);
        IntegerModifier shootDelayAfterReload = data.of("Shoot_Delay_After_Reload").serialize(IntegerModifier.class);

        return new ReloadModifier(magazineSize, ammoPerReload, reloadDuration, shootDelayAfterReload);
    }
}

File: src/main/kotlin/me/deecaad/weaponmechanicsplus/weapon/modifiers/ReloadModifier.kt
package me.deecaad.weaponmechanicsplus.weapon.modifiers

import me.deecaad.core.file.*
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.IntegerModifier
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.MechanicsModifier
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.MechanicsModifier.Companion.serializeMechanicsModifier

class ReloadModifier : Serializer<ReloadModifier> {

    var magazineSize: IntegerModifier? = null
    var ammoPerReload: IntegerModifier? = null
    var reloadDuration: IntegerModifier? = null
    var shootDelayAfterReload: IntegerModifier? = null
    var mechanicsModifier: MechanicsModifier? = null

    /**
     * Default constructor for serializer
     */
    constructor()

    constructor(magazineSize: IntegerModifier?, ammoPerReload: IntegerModifier?, reloadDuration: IntegerModifier?,
                shootDelayAfterReload: IntegerModifier?, mechanicsModifier: MechanicsModifier?) {
        this.magazineSize = magazineSize
        this.ammoPerReload = ammoPerReload
        this.reloadDuration = reloadDuration
        this.shootDelayAfterReload = shootDelayAfterReload
        this.mechanicsModifier = mechanicsModifier
    }

    @Throws(SerializerException::class)
    override fun serialize(data: SerializeData): ReloadModifier {
        val magazineSize = data.of("Magazine_Size").serialize(IntegerModifier::class.java)
        val ammoPerReload = data.of("Ammo_Per_Reload").serialize(IntegerModifier::class.java)
        val reloadDuration = data.of("Reload_Duration").serialize(IntegerModifier::class.java)
        val shootDelayAfterReload = data.of("Shoot_Delay_After_Reload").serialize(IntegerModifier::class.java)
        val mechanicsModifier = data.serializeMechanicsModifier()

        return ReloadModifier(magazineSize, ammoPerReload, reloadDuration, shootDelayAfterReload, mechanicsModifier)
    }
}